### PR TITLE
DSK-19772: Add the Dropbox example add-in to public add-ins repo

### PR DESCRIPTION
**Pull request by abhimanyusirohi**
This adds the `ChemDrawBox` add-in to our public add-ins repository. We need to provide an example add-in to the customer to demonstrate how the callback API works. This add-in shows the API in action using Dropbox as provider.

The add-in was written in React/TypeScript. It is the same add-in that was demoed in the last review meeting.

@PerkinElmer/chemdraw-desktop-reviewers

### CODE REVIEW

File: Examples/ChemDrawBox/src/registerServiceWorker.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+// In production, we register a service worker to serve assets from local cache.
+
+// This lets the app load faster on subsequent visits in production, and gives
+// it offline capabilities. However, it also means that developers (and users)
+// will only see deployed updates on the 'N+1' visit to a page, since previously
+// cached resources are updated in the background.
+
+// To learn more about the benefits of this model, read https://goo.gl/KwvDNy.
+// This link also includes instructions on opting out of this behavior.
+
```
Comment (stevebaxter) [inline Examples/ChemDrawBox/src/registerServiceWorker.ts]: We should remove this as it won't work in IE (which is what we are using for add-ins).

File: Examples/ChemDrawBox/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    "module": "esnext",
+    "target": "es5",
+    "lib": ["es6", "dom"],
+    "sourceMap": true,
+    "allowJs": true,
+    "jsx": "react",
+    "moduleResolution": "node",
+    "rootDir": "src",
+    "forceConsistentCasingInFileNames": true,
+    "noImplicitReturns": true,
```
Comment (stevebaxter) [inline Examples/ChemDrawBox/tsconfig.json]: Consider adding noUnusedLocals, see https://github.com/PerkinElmer/ChemDraw-Web/blob/master/tsconfig.json

File: Examples/ChemDrawBox/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+To install all required modules run the following command  
+`npm install` 
+
+To build the add-in, run:  
+`npm run build`
+
+> Note that OAuth requires a client-id which in the Dropbox's case is the _App-Key_ of the Dropbox app that authorizes. (See App Console here https://www.dropbox.com/developers). This _App-Key_ needs to be set in the application by replacing **YOUR_CLIENT_ID** in code.
+
+Once the build completes successfully, the _build_ folder will contain an _index.html_ file and a _static_ folder that contains the compiled CSS and JavaScript code. In order to create an add-in bundle, the following needs to be zipped into one archive:
```
Comment (stevebaxter) [inline Examples/ChemDrawBox/README.md]: Consider adding a script to package.json that does this.

**Comment by abhimanyusirohi** [comment]
> should we call this ChemDrawDropbox? Box and Dropbox are two very different products, both used at PKI

I wanted to name it `ChemDropbox` but the Dropbox's naming rules do not allow a Dropbox app name to have the word `Drop` in it so I carried the same rule here i.e. not using the word `Dropbox`.

File: Examples/ChemDrawBox/src/common/DropboxAPI.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+   }
+
+   /**
+    * Get user's files as JSON
+    */
+   public getUserFiles(): Promise<string> {
+        return new Promise<string>((resolve, reject) => {
+            if (this.accessToken.length === 0) {
+                reject(this.accessTokenErrorMessage);
+            }
```
[the remark this answers is not in the document]
Comment (abhimanyusirohi) [reply Examples/ChemDrawBox/src/common/DropboxAPI.ts]: Sure

**Comment by stevebaxter** [comment]
We need to find a different name then - Box is a different product. How about ChemDrawOAuthExample?

File: Examples/ChemDrawBox/README.md (hunk 3 above)
Comment (abhimanyusirohi) [reply Examples/ChemDrawBox/README.md]: Is there an example somewhere in our code that does something similar?

**Comment by abhimanyusirohi** [comment]
`ChemDrawOAuthExample` is good

**Comment by stevebaxter** [comment]
You could explain in the readme that this is an example that shows how to authorise to an application using OAuth, with Dropbox used as an example IdP?

File: Examples/ChemDrawBox/tsconfig.json (hunk 2 above)
Comment (abhimanyusirohi) [reply Examples/ChemDrawBox/tsconfig.json]: It exists, at the bottom

**Comment by abhimanyusirohi** [comment]
The latest change is to allow users to specify their own Dropbox app-key instead of hardcoding in code:

dropbox-addin

File: Examples/ChemDrawOAuthExample/src/components/UserInfo.tsx [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+              {fileArray}
+            </div>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  /**
+   * Called when one of the file button is clicked
```
Comment (JackTannerHotmail) [inline Examples/ChemDrawOAuthExample/src/components/UserInfo.tsx]: Minor typo: button -> buttons

**Comment by abhimanyusirohi** [comment]
@stevebaxter please review. I have made some slight modification to the add-in so the customer can use it right away without compiling their own. Also fixed some issues raised here, except for creating the script to bundle everything into a ChemDraw add-in.

File: Examples/ChemDrawOAuthExample/package.json [hunk 6]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "chemdropbox",
```
Comment (zhanglin-wu) [inline Examples/ChemDrawOAuthExample/package.json]: `chemdropbox` -> `ChemDrawOAuthExample`?

File: Examples/ChemDrawOAuthExample/public/manifest.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+{
+  "short_name": "React App",
```
Comment (zhanglin-wu) [inline Examples/ChemDrawOAuthExample/public/manifest.json]: `React App` -> `ChemDraw Add-in`?

File: Examples/ChemDrawOAuthExample/public/manifest.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  "name": "Create React App Sample",
```
Comment (zhanglin-wu) [inline Examples/ChemDrawOAuthExample/public/manifest.json]: `Create React App Sample` -> `ChemDraw Example OAuth Add-in`?

**Comment by zhanglin-wu** [comment]
Please add this comment in all sources files if possible:
```
```

**Comment by zhanglin-wu** [comment]
The following things are missing:
1. A description and a picture of the add-in in the introduction page of the examples:
    https://github.com/PerkinElmer/ChemDraw-AddIns/blob/master/Examples/README.md

2. An add-in like `ChemDrawOAuthExample.chemdrawaddin` that the the user can download and install into ChemDraw. We've provided such add-ins for all other examples.

File: Examples/ChemDrawOAuthExample/package.json (hunk 6 above)
Comment (abhimanyusirohi) [reply Examples/ChemDrawOAuthExample/package.json]: Changed, Thanks

File: Examples/ChemDrawOAuthExample/public/manifest.json (hunk 7 above)
Comment (abhimanyusirohi) [reply Examples/ChemDrawOAuthExample/public/manifest.json]: Done

File: Examples/ChemDrawOAuthExample/public/manifest.json (hunk 8 above)
Comment (abhimanyusirohi) [reply Examples/ChemDrawOAuthExample/public/manifest.json]: Done

**Comment by abhimanyusirohi** [comment]
1. Added the add-in to readme
2. Added a packaged version of the add-in
3. Added the file copyright comments to all code files

**Pull request merged**